### ISSUE TITLE
Sentences Couldn't Be Embedded by `embed.sh` (Encoder: 0 sentences in 0s)

### ISSUE THREAD

**Issue by ShunchiZhang** [facebookresearch/LASER#200]
I follow the instruction in 

https://github.com/facebookresearch/LASER/blob/7e60ad2d58a2ac4f0bb722d5cbfbe2dbf584c561/tasks/embed/README.md?plain=1#L26-L29

to embed `example.eng` to `example.eng.emb`:

```bash
(LASER) shunchiz@ubuntu:~/LASER/tasks/embed$ ./embed.sh example.eng example.eng.emb
couldn't find /home/shunchiz/LASER/models/wmt22/laser3-eng.v1.pt. defaulting to laser2
spm not specified, attempting default to: /home/shunchiz/LASER/models/wmt22/laser2.spm
 - Encoder: loading /home/shunchiz/LASER/models/wmt22/laser2.pt
 - SPM: processing example.eng  
 - Encoder: spm to example.eng.emb
 - Encoder: 0 sentences in 0s
```

It runs successfully, but in fact, it doesn't encode any sentences at all (`example.eng` is not empty).

How to resolve this issue? Any suggestions are highly appreciated :pray:!

---

**UPD:** I guess the models are well-downloaded as well.

```bash
(LASER) shunchiz@ubuntu:~/LASER/models$ find . -type f | xargs du | sort -nr
593840  ./wmt22/laser3-yor.v1.pt
593840  ./wmt22/laser3-wol.v1.pt
593828  ./wmt22/laser3-zul.v1.pt
593828  ./wmt22/laser3-xho.v1.pt
593828  ./wmt22/laser3-umb.v1.pt
593828  ./wmt22/laser3-tso.v1.pt
593828  ./wmt22/laser3-tsn.v1.pt
593828  ./wmt22/laser3-swh.v1.pt
593828  ./wmt22/laser3-ssw.v1.pt
593828  ./wmt22/laser3-som.v1.pt
593828  ./wmt22/laser3-sna.v1.pt
593828  ./wmt22/laser3-orm.v1.pt
593828  ./wmt22/laser3-nya.v1.pt
593828  ./wmt22/laser3-nso.v1.pt
593828  ./wmt22/laser3-luo.v1.pt
593828  ./wmt22/laser3-lug.v1.pt
593828  ./wmt22/laser3-lin.v1.pt
593828  ./wmt22/laser3-kin.v1.pt
593828  ./wmt22/laser3-kam.v1.pt
593828  ./wmt22/laser3-ibo.v1.pt
593828  ./wmt22/laser3-hau.v1.pt
593828  ./wmt22/laser3-fuv.v1.pt
449824  ./wmt22/laser3-amh.v1.pt
205216  ./bilstm.93langs.2018-12-26.pt
175236  ./wmt22/laser2.pt
163128  ./bilstm.eparl21.2018-11-19.pt
1124    ./wmt22/laser3-fuv.v1.spm
988     ./wmt22/laser2.spm
868     ./93langs.fvocab
808     ./93langs.fcodes
628     ./eparl21.fcodes
600     ./wmt22/laser3-fuv.v1.cvocab
524     ./eparl21.fvocab
480     ./wmt22/laser3-amh.v1.spm
460     ./wmt22/laser2.cvocab
168     ./wmt22/laser3-amh.v1.cvocab
```

**Comment by ShunchiZhang** [comment]
Almost the same error when explicitly specified one of the languages in WMT 22 - the only difference is: the model file `.pt` for laser3 can be found in the folder `$LASER/models/wmt22`.

https://github.com/facebookresearch/LASER/blob/7e60ad2d58a2ac4f0bb722d5cbfbe2dbf584c561/tasks/embed/README.md?plain=1#L31-L40

```bash
(LASER) shunchiz@ubuntu:/home/shunchiz/LASER/tasks/embed$ ./embed.sh example.tsn example.tsn.emb tsn
spm not specified, attempting default to: /home/shunchiz/LASER/models/wmt22/laser3-tsn.v1.spm
couldn't find laser3-tsn.v1.spm, defaulting to laser2.spm
 - Encoder: loading /home/shunchiz/LASER/models/wmt22/laser3-tsn.v1.pt
 - SPM: processing example.eng  
 - Encoder: spm to example.eng.emb
 - Encoder: 0 sentences in 0s
```

**Comment by wimmerth** [comment]
I have the same issue. Can anyone help?

**Comment by ShunchiZhang** [comment]
# Solution

# TL;DR:

I ran a debugger and found the solution: simply `pip install transliterate`.

## Locate the error

Though they say `transliterate` package is only used for Greek

https://github.com/facebookresearch/LASER/blob/f4689ad6333a8bed85b3c975b1c57e745e3e5623/README.md?plain=1#L39

but in `LASER/source/lib/romanize_lc.py` (which would be called necessarily, if you execute this `LASER/tasks/embed/embed.sh` script)

https://github.com/facebookresearch/LASER/blob/f4689ad6333a8bed85b3c975b1c57e745e3e5623/source/lib/romanize_lc.py#L20

it still needs `transliterate` package!

## Why not raise an error?

The reason why it doesn't raise `ModuleNotFoundError` is because this `romanize_lc.py` is called by `subprocess.check_output()` function. In other words, this `romanize_lc.py` is executed through a bash subprocess.

You can look into `SPMApply()` function in `LASER/source/lib/text_processing.py` for details.

https://github.com/facebookresearch/LASER/blob/f4689ad6333a8bed85b3c975b1c57e745e3e5623/source/lib/text_processing.py#L131-L138

where `ROMAN_LC` in line 135 above is defined as

https://github.com/facebookresearch/LASER/blob/f4689ad6333a8bed85b3c975b1c57e745e3e5623/source/lib/text_processing.py#L39

**Comment by ShunchiZhang** [comment]
> I have the same issue. Can anyone help?

Hi @wimmerth, I just found and updated the solution. Take a look and I hope it helps :)

**Comment by wimmerth** [comment]
Works like a charm now, thank you!
